client/src/components/RecipeSearch.js

```diff
@@ -0,0 +1,142 @@
+import React, { useState } from "react"
+import SearchedRecipeTile from "./SearchedRecipeTile"
+
+const RecipeSearch = (props) => {
+  const [formData, setFormData] = useState({
+    searchQuery: ""
+  })
+
+  const handleChange = (event) => {
+    setFormData({
+      ...formData,
+      [event.currentTarget.name]: event.currentTarget.value
+    })
+  }
+  const [searchResults, setSearchResults] = useState(undefined)
+
+  const handleSubmit = async (event) => {
+    event.preventDefault()
+    try {
+      const response = await fetch("/api/v1/recipe-search", {
+        method: "POST",
+        headers: new Headers({
+          "Content-Type": "application/json",
+        }),
+        body: JSON.stringify(formData)
+      })
+      if(response.ok) {
+        const body = await response.json()
+        setSearchResults(body.searchResults)
+      }
+    } catch(err) {
+      console.error(`Error in fetch: ${err.message}`)
+    }
+  }
+
+  const handlePageClick = async (event) => {
+    event.preventDefault()
+    const targetPage = event.currentTarget.getAttribute("value")
+    const pageData = {
+      targetPage,
+      searchQuery: searchResults.searchQuery
+    }
+    try {
+      const response = await fetch("/api/v1/recipe-search/next", {
+        method: "POST",
+        headers: new Headers({
+          "Content-Type": "application/json"
+        }),
+        body: JSON.stringify(pageData)
+      })
+      if(response.ok) {
+        const body = await response.json()
+        setSearchResults(body.searchResults)
+      }
+    } catch(err) {
+      console.error(`Error in fetch: ${err.message}`)
+    }
+  }
+
+  let searchList
+  if(searchResults) {
+    searchList = searchResults.results.map(result => {
+      return(
+        <SearchedRecipeTile
+          key={result.id} 
+          result={result}
+        />
+      )
+    })
+  }
+
+  let pagesDisplay
+  if(searchResults) {
+    const totalPages  = Math.ceil(searchResults.totalResults / 10)
+    const currentPage = searchResults.offset/10
+    let pageNums = []
+    for (let x = 0; x < totalPages; x++) {
+      if(x === currentPage) {
+        pageNums.push(x+1)
+      } else if(x >= currentPage - 3 && x <= currentPage + 3) {
+        pageNums.push(
+          <a
+            key={x}
+            value={x}
+            onClick={handlePageClick}
+          >
+          {x+1}
+          </a>
+        )
+      }
+    }
```

Comment:
while this logic is awesome, the cleanest method would be: only manage 10 results in state at a time (rather than ALL of the potentially hundreds of matching searched recipes). Then whenever they click on "2", you make a request to the backend that uses the offset method to retrieve the next 10 * the number clicked

the less logic we can have in React the better, because even though the additional fetch request will take time, computations in react rely on the active memory of some persons browser. If they have an old browser on an old phone, this will make the website sluggish. Its a wild edge case, but something to consider

https://knexjs.org/#Builder-offset

Reply:
Spoonacular actually only returns 10 results at a time, so I'll never have more than 10 recipes to store in state.  This is the implementation I used, that each click on "previous", "next", or a page number will send a new fetch request that uses the targetPage*10 as an offset.